src/tensor_ops/matmul.rs

```diff
@@ -122,131 +141,279 @@ pub fn vecmat_mul<const N: usize, const O: usize, H: Tape>(
+/// matrix multiply `c += a * trans(b)`
+fn mm_bt<const M: usize, const K: usize, const N: usize>(
+    a: &[[f32; K]; M],
+    b_t: &[[f32; K]; N],
+    c: &mut [[f32; N]; M],
 ) {
+    let a = a.as_ptr() as *const f32;
+    let b_t = b_t.as_ptr() as *const f32;
+    let c = c.as_mut_ptr() as *mut f32;
+
+    #[cfg(not(feature = "intel-mkl"))]
+    unsafe {
+        matrixmultiply::sgemm(
+            M, K, N, 1.0, a, K as isize, 1, b_t, 1, K as isize, 1.0, c, N as isize, 1,
+        )
+    }
+
+    #[cfg(feature = "intel-mkl")]
     unsafe {
-        let a = x.as_ptr() as *const f32;
-        let b = y_t.as_ptr() as *const f32;
-        let c = out.as_mut_ptr() as *mut f32;
-        sgemm(
-            M, N, O, 1.0, a, N as isize, 1, b, 1, N as isize, 1.0, c, O as isize, 1,
+        cblas_sys::cblas_sgemm(
+            cblas_sys::CblasRowMajor,
+            cblas_sys::CblasNoTrans,
+            cblas_sys::CblasTrans,
+            M as libc::c_int,
+            N as libc::c_int,
+            K as libc::c_int,
+            1.0,
+            a,
+            K as libc::c_int,
+            b_t,
+            K as libc::c_int,
+            1.0,
+            c,
+            N as libc::c_int,
         )
-    };
+    }
 }
 
-/// matrix multiply `transpose(out) = transpose(x) * y + beta * transpose(out)`
-fn matmat_mul_into_xtzt<const M: usize, const N: usize, const O: usize>(
-    x_t: &[[f32; M]; N],
-    y: &[[f32; O]; N],
-    out_t: &mut [[f32; M]; O],
+/// matrix multiply `trans(c) += trans(a) * b`
+fn mm_atct<const M: usize, const K: usize, const N: usize>(
+    a_t: &[[f32; M]; K],
+    b: &[[f32; N]; K],
+    c_t: &mut [[f32; M]; N],
 ) {
+    let a_t = a_t.as_ptr() as *const f32;
+    let b = b.as_ptr() as *const f32;
+    let c_t = c_t.as_mut_ptr() as *mut f32;
+
+    #[cfg(not(feature = "intel-mkl"))]
     unsafe {
-        let a = x_t.as_ptr() as *const f32;
-        let b = y.as_ptr() as *const f32;
-        let c = out_t.as_mut_ptr() as *mut f32;
-        sgemm(
-            M, N, O, 1.0, a, 1, M as isize, b, O as isize, 1, 1.0, c, 1, M as isize,
+        matrixmultiply::sgemm(
+            M, K, N, 1.0, a_t, 1, M as isize, b, N as isize, 1, 1.0, c_t, 1, M as isize,
         )
-    };
+    }
+
+    #[cfg(feature = "intel-mkl")]
+    unsafe {
+        cblas_sys::cblas_sgemm(
+            cblas_sys::CblasColMajor,
+            cblas_sys::CblasNoTrans,
+            cblas_sys::CblasTrans,
+            M as libc::c_int,
+            N as libc::c_int,
+            K as libc::c_int,
+            1.0,
+            a_t,
+            M as libc::c_int,
+            b,
+            N as libc::c_int,
+            1.0,
+            c_t,
+            M as libc::c_int,
+        )
+    }
 }
 
-fn vecmat_mul_into<const N: usize, const O: usize>(
-    x: &[f32; N],
-    y: &[[f32; O]; N],
-    out: &mut [f32; O],
-) {
+/// vector matrix multiply `c += a * b`
+fn vm<const K: usize, const N: usize>(a: &[f32; K], b: &[[f32; N]; K], c: &mut [f32; N]) {
```

Comment:
This and other vector functions should use whatever the mv functions are instead of using sgemm